kubetest/azure.go

```diff
@@ -0,0 +1,348 @@
+		ctx:            context.Background(),
+		clientId:       ClientID,
+		clientSecret:   ClientSecret,
+		tenantId:       TenantId,
+		subscriptionId: SubscriptionId,
+		apiModelPath:   *acsTemplatePath,
+		name:           *acsResourceName,
+		dnsPrefix:      *acsDnsPrefix,
+		location:       *acsLocation,
+		resourceGroup:  *acsResourceGroupName,
+		outputDir:      tempdir,
+		sshPublicKey:   fmt.Sprintf("%s", sshKey),
+	}, nil
+}
+
+func (c *Cluster) generateTemplate() error {
+	v := &AcsEngineApiModel{
+		ApiVersion: "vlabs",
+		Location:   c.location,
+		Name:       c.name,
+		Tags: map[string]string{
+			"date": time.Now().String(),
+		},
+		Properties: &Properties{
+			OrchestratorProfile: &OrchestratorProfile{
+				OrchestratorType: "Kubernetes",
+			},
+			MasterProfile: &MasterProfile{
+				Count:     1,
+				DNSPrefix: c.dnsPrefix,
+				VMSize:    *acsMasterVmSize,
+			},
+			AgentPoolProfiles: []*AgentPoolProfile{
+				{
+					Name:                "agentpool0",
+					VMSize:              *acsAgentVmSize,
+					Count:               *acsAgentPoolCount,
+					OSType:              "Windows",
+					AvailabilityProfile: "AvailabilitySet",
+				},
+			},
+			LinuxProfile: &LinuxProfile{
+				AdminUsername: *acsAdminUsername,
+				SSHKeys: &SSH{
+					PublicKeys: []PublicKey{{
+						KeyData: c.sshPublicKey,
+					},
+					},
+				},
+			},
+			WindowsProfile: &WindowsProfile{
+				AdminUsername: *acsAdminUsername,
+				AdminPassword: *acsAdminPassword,
+			},
+			ServicePrincipalProfile: &ServicePrincipalProfile{
+				ClientID: ClientID,
+				Secret:   ClientSecret,
+			},
+		},
+	}
+	apiModel, _ := json.Marshal(v)
+	c.apiModelPath = path.Join(c.outputDir, "kubernetes.json")
+	err := ioutil.WriteFile(c.apiModelPath, apiModel, 0644)
+	if err != nil {
+		return fmt.Errorf("Cannot write to file: %v", err)
+	}
+	return nil
+}
+
+func (c *Cluster) getAcsEngine(retry int) error {
+	downloadPath := path.Join(os.Getenv("HOME"), "acs-engine")
+	f, err := os.Create(downloadPath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	for i := 0; i < retry; i++ {
+		log.Printf("downloading %v from %v", downloadPath, *acsEngineURL)
+		if err := httpRead(*acsEngineURL, f); err == nil {
+			break
+		}
+		err = fmt.Errorf("url=%s failed get %v: %v", *acsEngineURL, downloadPath, err)
+		if i == retry-1 {
+			return err
+		}
+		log.Println(err)
+		sleep(time.Duration(i) * time.Second)
+	}
+
+	f.Close()
+	o, err := control.Output(exec.Command("md5sum", f.Name()))
+	if err != nil {
+		return err
+	}
+	if string(o) != *acsEngineMD5 {
+		return fmt.Errorf("Wrong md5 sum for acs-engine")
+	}
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		return fmt.Errorf("unable to get current directory: %v", err)
+	}
+	log.Printf("Extracting tar file %v into directory %v", f.Name(), cwd)
+
+	if err = control.FinishRunning(exec.Command("tar", "-xzf", f.Name(), "--strip", "1")); err != nil {
+		return err
+	}
+	c.acsEngineBinaryPath = path.Join(cwd, "acs-engine")
```

Comment:
Check whether file exists first, download only if not preset. This is also useful for testing